processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -1486,6 +1454,30 @@ public int getKeyBufferValueSize()
       }
 
       @Override
+      public void getFromByteBuffer(ByteBuffer buffer, int initialOffset, int dimValIdx, Comparable[] dimValues)
+      {
+        dimValues[dimValIdx] = dictionary.get(buffer.getInt(initialOffset + keyBufferPosition));
+      }
+
+      @Override
+      public int compare(ByteBuffer lhsBuffer, ByteBuffer rhsBuffer, int lhsPosition, int rhsPosition)
+      {
+        return bufferComparator.compare(lhsBuffer, rhsBuffer, lhsPosition, rhsPosition);
```

Comment:
Couldn't this indirection be reasonably avoided here and in other helpers?

Reply:
I guess it's difficult. The indirection here is needed because we are using dictionary encoding for string type. To avoid this, we need to compare strings directly, but I'm not sure it is better than this.

Reply:
I meant indirection for the use of `helper.compare()`, which is delegated to `bufferComparator`. It seems that comparator is always created, so it could be exposed, and the called could avoid this indirection

Reply:
You mean like this?

```java
      @Override
      public int compare(ByteBuffer lhsBuffer, ByteBuffer rhsBuffer, int lhsPosition, int rhsPosition)
      {
        if (!pushLimitDown) {
          return Ints.compare(
              rankOfDictionaryIds[lhsBuffer.getInt(lhsPosition + keyBufferPosition)],
              rankOfDictionaryIds[rhsBuffer.getInt(rhsPosition + keyBufferPosition)]
          );
        } else {
          String lhsStr = dictionary.get(lhsBuffer.getInt(lhsPosition + keyBufferPosition));
          String rhsStr = dictionary.get(rhsBuffer.getInt(rhsPosition + keyBufferPosition));
          return realComparator.compare(lhsStr, rhsStr);
        }
      }
```

Reply:
No I meant that Helper has a method like getComparator(), then those comparators collected into an array and accessed directly in a hot spot method

Reply:
I got your point, but it's already done in that way. KeySerde.bufferComparator() is called one time in iterator().

Reply:
I don't see. The line to which this thread is attached is on `GrouperHelper.compare()`, that delegates to `bufferComparator` field. I don't see this field is exposed, but `GrouperHelper.compare()` is called in some hot spots.

Reply:
Ok. Exposed bufferComparators.